pm: policy: add support for events

gmarull opened the pull request:
Events in the power-management policy context are defined as any source that will wake up the system at a known time in the future. By registering such event, the policy manager will be able to decide wether certain power states are worth entering or not.

Events will bypass the ticks argument received by the policy manager if they occur earlier.

gmarull committed:
pm: policy: add support for events

Events in the power-management policy context are defined as any source
that will wake up the system at a known time in the future. By
registering such event, the policy manager will be able to decide wether
certain power states are worth entering or not.

Events will bypass the ticks argument received by the policy manager if
they occur earlier.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

hubertmis commented on include/zephyr/pm/policy.h:
```diff
@@ -158,6 +165,43 @@ void pm_policy_latency_changed_subscribe(struct pm_policy_latency_subscription *
+ * event, the policy manager will be able to decide wether certain power states
+ * are worth entering or not.
+ *
+ * @note Expired events need to be unregistered manually using
```
It would be nice to be able to remove this responsibility from the API user. But it can be enhanced later if we find it beneficial

gmarull replied:
It's not easy because the PM subsystem doesn't have any link with such events. It just knows _when_ it will happen.

ceolin replied:
What happens if they are not removed ? They eventually be accounted again and in a not intuitive way since it will depend on if a newer event is added or not.

How this is supposed to be used ? My main concern is how it is expected to the user know when to remove it since there is not feedback about when it expired, they should set a timeout ? if so the timeout itself would be accounted by the policy.

What about you register three events `a (now() + 100)`,  `b (now() + 150)`,  `c(now() + 175)`;
It is expected that the policy understands that there will be events within 100 us then another one 50us later and after that one 25us later, but if the no event is removed / added or updated, `next_event_cyc` will account only for `a` . Am I missing something ?

hubertmis reviewed (commented):
I like the feature of registering other wake-up events than OS timer to properly select the next power state.

I think we are missing some code around https://github.com/zephyrproject-rtos/zephyr/blob/933d338c9783cce55466fed000baf2f3beeac891/subsys/pm/pm.c#L239 that would allow to remove latency for such events.

gmarull commented:
> I like the feature of registering other wake-up events than OS timer to properly select the next power state.
> 
> I think we are missing some code around
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/933d338c9783cce55466fed000baf2f3beeac891/subsys/pm/pm.c#L239

Why should we modify this? The scheduled event is expected to wake up the system, right?
> 
> that would allow to remove latency for such events.

hubertmis commented:
> Why should we modify this? The scheduled event is expected to wake up the system, right?

Let's say the expected event is the end of ongoing SPI transaction. It is expected to arrive at time X. The event will wake up the CPU/system/whatever-is-sleeping. The wake-up procedure duration is D. CPU starts handling the event associated with the SPI transaction at Y = X + D.

The SPI end transaction handler might have a requirement to handle the event within acceptable latency A, so the event handler must be executed by CPU no later than Z = X + A.

If D > A we have a problem, because the event is handled at Y that happens later than Z (the acceptable latency).

To fix the problem, the power management subsystem can schedule a CPU wake-up at W = X - D. If the wake-up is triggered at W and its duration is D, then the system is up and ready to handle the event at X with zero latency. In such circumstances we don't need to worry about A, Y, or Z timings.

Alternatively, the `pm_policy_event_register()` could take one more argument describing if the event tolerates latency and set the CPU wake-up timer at W only for events that do not tolerate latency.

gmarull commented:
> > Why should we modify this? The scheduled event is expected to wake up the system, right?
> 
> Let's say the expected event is the end of ongoing SPI transaction. It is expected to arrive at time X. The event will wake up the CPU/system/whatever-is-sleeping. The wake-up procedure duration is D. CPU starts handling the event associated with the SPI transaction at Y = X + D.
> 
> The SPI end transaction handler might have a requirement to handle the event within acceptable latency A, so the event handler must be executed by CPU no later than Z = X + A.
> 
> If D > A we have a problem, because the event is handled at Y that happens later than Z (the acceptable latency).
> 
> To fix the problem, the power management subsystem can schedule a CPU wake-up at W = X - D. If the wake-up is triggered at W and its duration is D, then the system is up and ready to handle the event at X with zero latency. In such circumstances we don't need to worry about A, Y, or Z timings.
> 
> Alternatively, the `pm_policy_event_register()` could take one more argument describing if the event tolerates latency and set the CPU wake-up timer at W only for events that do not tolerate latency.

I see the point, however, the `sys_clock_set_timeout` resolution is again in ticks (coarse). I'm not sure if it would work. The other solution would be to impose a latency requirement by the driver, so that system is always kept in a sleep state that allows for fast enough wakeup. This should work if the event is expected to happen _quickly_, otherwise we'd be wasting power by not going to deeper sleep states.

hubertmis commented:
> the `sys_clock_set_timeout` resolution is again in ticks (coarse)

I think that's acceptable. When we decide to create more fine-grained latency control, we can modify `sys_clock_set_timeout` for both system timer and external events.

> otherwise we'd be wasting power by not going to deeper sleep states

That's exactly the scenario I would like to prevent: so that users of this API, when they register events they do not mix it with latency management functions. Latency management functions should be used only for events that happen at unpredictable time and cannot be registered.

hubertmis reviewed (dismissed):
It looks ready for the initial solution.
I hope the discussed improvements like the latency control will be added later.

ceolin commented on include/zephyr/pm/policy.h:
```diff
@@ -158,6 +165,43 @@ void pm_policy_latency_changed_subscribe(struct pm_policy_latency_subscription *
+ *
+ * @see pm_policy_event_unregister
+ */
+void pm_policy_event_register(struct pm_policy_event *evt, uint32_t time);
```
I think it is worth to change the variable to `time_us` just to make ti clear the resolution.

gmarull replied:
I think that makes sense, however, I'd also adjust other calls like the latency ones (which use `time` for the same)

ceolin commented on include/zephyr/pm/policy.h:
```diff
@@ -48,6 +48,13 @@ struct pm_policy_latency_request {
+struct pm_policy_event {
+	sys_snode_t node;
+	/** Request value. */
+	uint32_t value;
```
value is quite confuse IMHO. Can we make it more explicit like `cycs_to_event` or something that says what it is representing ?

gmarull replied:
Note this is an internal value, but we can maybe adjust its name explicitly mentioning this is for internal use.

ceolin reviewed (changes requested):
Giving - 1 to avoid merging it while we discuss the API usage. It is a little bit confuse about how to use it IMHO. Specially around what is the behavior when the event is not removed.

gmarull commented:
> Giving - 1 to avoid merging it while we discuss the API usage. It is a little bit confuse about how to use it IMHO. Specially around what is the behavior when the event is not removed.

So the idea here is that once the event happens, you're responsible to unregister the event. Not doing so would be an API contract violation. These events are just informing the PM subsys, but there's no strict linking between these events <> PM subsys. In practice, what would happen is that the event would fall behind the cycle counter, and so considered to be an event _far_ in the future (after cycle counter rollover). So if not removed it would periodically kick in as cycle counter approaches again. I will try to better document that removing the event after it happens is a mandatory thing to do.

gmarull committed:
pm: policy: clarify latency units and clarify private members

Latency is always specified in microseconds, so apend `us` to the
relevant variables. Also, make it clear that latency request structures
are private (they are just exposed publicly to allow static allocation).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: boot: mcuboot: disable test

Disable test due to observed issues with CONFIG_BUILD_OUTPUT_ADJUST_LMA.
Refer to https://github.com/zephyrproject-rtos/zephyr/issues/58080.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

ceolin commented:
Thanks for updating the documentation, it is clear now how to use it. Question, when this is supposed to be used ? Which sort of event you need to register and how to you now that the event has already happen ?

hubertmis commented:
> Which sort of event you need to register and how to you now that the event has already happen ?

It is mostly about IRQs which happen at predictable times. E.g. when UART is transferring out a "Hello world" message the UART driver configures UART's DMA, starts the transfer, and waits on a semaphore for a TX_FINISHED IRQ. While the driver waits, the idle thread is executed and runs the PM sequence, which can result in entering deep sleep, which would be immediately exited because of the TX_FINISHED IRQ. Immediately exiting deep sleep is not power efficient.

While the driver configures UART's DMA, it can predict when the UART's TX_FINISHED IRQ will happen, based on the baud rate and the transfer size. The driver is supposed to use the new API and register this event in the PM so that PM policy restrains from entering deep sleep, knowing an IRQ waking up the system will happen in e.g. 1 ms. In this scenario, the PM policy could select light sleep that would be more power efficient.

ceolin commented:
> > Which sort of event you need to register and how to you now that the event has already happen ?
> 
> It is mostly about IRQs which happen at predictable times. E.g. when UART is transferring out a "Hello world" message the UART driver configures UART's DMA, starts the transfer, and waits on a semaphore for a TX_FINISHED IRQ. While the driver waits, the idle thread is executed and runs the PM sequence, which can result in entering deep sleep, which would be immediately exited because of the TX_FINISHED IRQ. Immediately exiting deep sleep is not power efficient.
> 
> While the driver configures UART's DMA, it can predict when the UART's TX_FINISHED IRQ will happen, based on the baud rate and the transfer size. The driver is supposed to use the new API and register this event in the PM so that PM policy restrains from entering deep sleep, knowing an IRQ waking up the system will happen in e.g. 1 ms. In this scenario, the PM policy could select light sleep that would be more power efficient.
> 
> The UART driver can unregister the event while handling TX_FINISHED IRQ.

Thanks ! It makes sense.

gchwier commented on tests/boot/test_mcuboot/testcase.yaml:
```diff
@@ -12,6 +12,7 @@ common:
 tests:
   boot.mcuboot:
     tags: mcuboot
+    skip: true # disabled due to issue #58080
```
test below: boot.mcuboot.assert, also should be skipped
